### ISSUE TITLE
[java] Flag literals in all conditions

### ISSUE THREAD

**Issue by numeralnathan** [pmd/pmd#2287]
**Affects PMD Version:** 

6.21.0

**Rule:**

`AvoidLiteralsInIfCondition`

**Description:**

`AvoidLiteralsInIfCondition` flags literals in `if` conditions.  What about literals in `while` conditions, `for` conditions, ternary conditions or any condition?  Should literals be flagged in these situations?

If a rule does flag all of these situations, should the rule be renamed to `AvoidLiteralsInCondition`?

**Code Sample demonstrating the issue:**

The following is not valid Java code.  However, if pasted in PMD Designer, it can be used to develop the rule.

```
for (i = 0; i < 10; i++)   // 10 should be flagged
if (i == 0)   // 0 should be ignored
if (i == 1 + 5)   // 1 and 5 should be flagged
while (++i < 100)   // 100 should be flagged
j = (i < 1000) && (j > 7) ? i + 1 : i - 1;   // 1000 and 7 should be flagged
if (i > 20)   // 20 should be flagged
if (compareAndSet(this, index, index + 4))  // 4 should be ignored
if (m_indexes[m_indexes.length - 9487] >= m_size)  // 9487 should be ignored
return j > 8443;   // 8443 should be flagged
return (j > 15) && (i < 13);  // 15 and 13 should be flagged
return 17L << (ordinal & (Long.SIZE - 18));   // 17 and 18 should be ignored
```

**XPath:**

I looked at the XPath in `AvoidLiteralsInIfCondition` and produced this scant XPath.  It lacks conditions on the parameter values.

```
<!-- Handles literals in if, for and while statements -->
(//IfStatement | //ForStatement | //WhileStatement)
/Expression
//*
/PrimaryExpression
/PrimaryPrefix
/Literal
[
   not(NullLiteral) and
   not(BooleanLiteral) and
   not(ancestor::ArgumentList) and
   not(ancestor::PrimarySuffix[@ArrayDereference = "true"]) and
   not(@Image = "0") and
   @StringLiteral = "false"
]
|
<!-- Handles literals in return statements if there is a relational expression
//Expression
/RelationalExpression
/PrimaryExpression
/PrimaryPrefix
/Literal
[
   not(NullLiteral) and
   not(BooleanLiteral) and
   not(ancestor::ArgumentList) and
   not(ancestor::PrimarySuffix[@ArrayDereference = "true"]) and
   ancestor::ReturnStatement and
   not(@Image = "0") and
   @StringLiteral = "false"
]
|
<!-- Handles literals in ternary operators -->
//ConditionalExpression
//RelationalExpression
/PrimaryExpression
/PrimaryPrefix
/Literal
[
   not(NullLiteral) and
   not(BooleanLiteral) and
   not(ancestor::ArgumentList) and
   not(ancestor::PrimarySuffix[@ArrayDereference = "true"]) and
   not(@Image = "0") and
   @StringLiteral = "false"
]						
```

**Running PMD through:** Eclipse

**Comment by adangel** [comment]
I think it makes sense to broaden the usage scenario of this rule - in which case we definitely need to create a new rule (like you suggested AvoidLiteralsInConditions) and deprecate the existing rule.

I don't know, why the rule only looks at if conditions - probably it was the one itch that has been solved at that time. Looking at your examples, I wonder whether e.g. `int a = 10` should be flagged as well? It's not a condition, but a literal - 10 has probably a meaning beyond being 10. So, it could even be a rule, to declare for each literal a meaningful constant name. This rule could surely be overwhelming - especially if run against an existing code base... 
A related rule is [AvoidDuplicateLiterals](https://pmd.github.io/latest/pmd_rules_java_errorprone.html#avoidduplicateliterals).

**Comment by numeralnathan** [comment]
> I wonder whether e.g. `int a = 10` should be flagged as well? It's not a condition, but a literal - 10 has probably a meaning beyond being 10.

I was wondering the same thing.  If we do create such a rule, it should have a property to specify which literals to ignore.  The literals `-1`, `0` and `1` may need to be ignored since they occur in many places where giving them a constant name may not help clarity.  See the following code.  Changing the `1` to a constant name such `PREVIOUS` may not be very clear.

```
List<?> list;
int size;

size = list.size();

list.get(size - 1);
```

**Comment by Fernal73** [comment]
Maybe, we're approaching this the wrong way. Perhaps, we ought to look at
when we would want or have to use literals.

https://sourceforge.net/p/pmd/discussion/188194/thread/336c5befc3/

**Comment by Fernal73** [comment]
Another point is with regards to the default values specified in the rule definition.
I see no reason why 1 should not be in there as well.

https://stackoverflow.com/a/63238205/3924108
